tokio-macros/src/entry.rs

```diff
@@ -389,19 +390,120 @@ fn build_config(
     config.build()
 }
 
+fn fn_without_args(mut input: ItemFn, is_test: bool, config: FinalConfig) -> TokenStream {
+    let async_keyword = input.sig.asyncness.take();
+    let fn_sig = &input.sig;
+    let fn_name = &input.sig.ident;
+
+    let crate_path = config
+        .crate_name
+        .map(ToTokens::into_token_stream)
+        .unwrap_or_else(|| Ident::new("tokio", Span::call_site()).into_token_stream());
+
+    let mut rt = match config.flavor {
+        RuntimeFlavor::CurrentThread => quote! {
+            #crate_path::runtime::Builder::new_current_thread()
+        },
+        RuntimeFlavor::Threaded => quote! {
+            #crate_path::runtime::Builder::new_multi_thread()
+        },
+    };
+    if let Some(v) = config.worker_threads {
+        rt = quote! { #rt.worker_threads(#v) };
+    }
+    if let Some(v) = config.start_paused {
+        rt = quote! { #rt.start_paused(#v) };
+    }
+    if let Some(v) = config.unhandled_panic {
+        let unhandled_panic = v.into_tokens(&crate_path);
+        rt = quote! { #rt.unhandled_panic(#unhandled_panic) };
+    }
+
+    let generated_attrs = if is_test {
+        quote! { #[::core::prelude::v1::test] }
+    } else {
+        quote! {}
+    };
+
+    // This explicit `return` is intentional. See tokio-rs/tokio#4636
+    let last_block = quote! {
+        #[allow(clippy::expect_used, clippy::diverging_sub_expression)]
+        #rt.enable_all().build().expect("Failed building the Runtime").block_on(body)
+    };
+
+    let body = input.body;
+
+    input.body = if is_test {
+        let output_type = match &input.sig.output {
+            syn::ReturnType::Default => quote! { () },
+            syn::ReturnType::Type(_, ret_type) => quote! { #ret_type },
+        };
+        quote! {
+            #async_keyword #fn_sig #body
+            let body = #fn_name();
+
+            #crate_path::pin!(body);
+            let body: ::core::pin::Pin<&mut dyn ::core::future::Future<Output = #output_type>> = body;
+        }
+    } else {
+        quote! {
+            #async_keyword #fn_sig #body
+            let body = #fn_name();
+        }
+    };
+
+    input.into_tokens(generated_attrs, last_block)
+}
+
+fn has_self_keyword(mut iter: token_stream::IntoIter) -> bool {
+    iter.any(|tt| match tt {
+        TokenTree::Ident(ident) => ident == "Self",
+        TokenTree::Group(group) => has_self_keyword(group.stream().into_iter()),
+        _ => false,
+    })
+}
+
 fn parse_knobs(mut input: ItemFn, is_test: bool, config: FinalConfig) -> TokenStream {
-    input.sig.asyncness = None;
+    if is_test || input.sig.inputs.is_empty() && !has_self_keyword(input.body.clone().into_iter()) {
```

Comment:
I think the cases where `Self` is included in the return type, generics, or where clause will still be missed.

Reply:
I should be fine. We don't need to fallback to async block, If generics or return position contain `Self` keyword.

Reply:
No. For example, the following conversion is incorrect:
```rs
impl MyStruct {
    // From:
    // #[tokio::main]
    // async fn func() -> Self { todo!() }
    //
    // To:
    fn func() -> Self {
        async fn func() -> Self { todo!() }
        let body = func();
        // ...
    }
}
```

```
error[E0401]: can't use `Self` from outer item
 --> src/lib.rs:9:28
  |
2 | impl MyStruct {
  | ---- `Self` type implicitly declared here, by this `impl`
...
9 |         async fn func() -> Self { todo!() }
  |                            ^^^^
  |                            |
  |                            use of `Self` from outer item
  |                            refer to the type directly here instead
```